source/simulator/core.cc

```diff
@@ -2241,6 +2256,297 @@ namespace aspect
           break;
         }
 
+        case NonlinearSolver::Newton_Stokes:
+        {
+          if (parameters.free_surface_enabled)
+            free_surface->execute ();
+
+          double initial_temperature_residual = 0;
+          double initial_stokes_residual      = 0;
+          std::vector<double> initial_composition_residual (parameters.n_compositional_fields,0);
+
+          //parameters.nonlinear_iteration_number = 0;
+
+          double max = std::numeric_limits<double>::max();
+
+          double initial_newton_residual_velo = 0;
+          double initial_newton_residual_pres = 0;
+          double initial_newton_residual = 0;
+
+          double newton_residual_velo = 0;
+          double newton_residual_pres = 0;
+          double newton_residual = 0;
+          double newton_residual_old = 0;
+
+          parameters.switch_initial_newton_residual = 1;
+          parameters.newton_residual = 1;
+          parameters.newton_theta = 0;
+
+          const double piccard_linear_stokes_solver_tolerance = parameters.linear_stokes_solver_tolerance;
+
+          bool   use_picard = true;
+
+          for (unsigned int i=0; (! ((i >= parameters.max_nonlinear_iterations) // regular timestep
+                                     ||
+                                     ((pre_refinement_step < parameters.initial_adaptive_refinement) // pre-refinement
```

Comment:
You can move a lot of this outside of the for-loop like following:
```
const unsigned int max_iterations = (pre_refinement_step < parameters.initial_adaptive_refinement)
?
std::min (parameters.max_nonlinear_iterations,parameters.max_nonlinear_iterations_in_prerefinement)
:
parameters.max_nonlinear_iterations;

for (unsigned int i=0; i<max_iterations;++i)
...
```
Please check that the logic actually does what it is supposed to do, but if that works it would be much easier to understand than the current form.